broadinstitute/single_cell_portal_core
# Complete backend integration for differential expression jobs (SCP-4195, SCP-4212)

This update expands on #1483 to now allow launching of differential expression jobs more seamlessly via the `IngestJob` and `DifferentialExpressionService` classes.  This new integration will handle all input validation, launch the requested job, poll for completion, and handle success/failure states.  This includes adding a new attribute `is_differential_expression_enabled` for `CellMetadatum` documents, and a similar flag on cluster-based cell annotations.  The above information will be surfaced in the study "Explore" info object, under the `annotationList` attribute (this contains a list of all available annotations, which will declare `is_differential_expression_enabled`).

This update also addresses a lingering issue with some test suites missing necessary includes that prevented them from being run locally.  This did not affect CI, but prevented some integration tests from being run with `rails test [test_path]`.

NOTE: This integration does not include deleting differential expression outputs if a user subsequently deletes metadata/cluster files.  Since this integration will only be an admin-only console command to launch, it is reasonable to assume admins can clean up files in study buckets if needed.

MANUAL TESTING
1. Pull this branch and run any pending migrations with `rake db:migrate`
2. If you have not already done so, please follow steps 1-4 from #1483 to set the correct Ingest Docker image and populate the required test data
3. Once the data is ingested, open a Rails console session and validate that the requested `CellMetadatum` object does not have `is_differential_expression_enabled` set yet, using the study accession of the above study:
```
study = Study.find_by(accession: <new study accession>)
metadata = study.cell_metadata.by_name_and_type('cell_type__ontology_label', 'group')
metadata.is_differential_expression_enabled
=> false
```
4. Now launch the DE job:
```

cluster_file = study.cluster_ordinations_files.first
user = study.user
options = {
  annotation_name: 'cell_type__ontology_label',
  annotation_type: 'group', 
  annotation_scope: 'study'
}
DifferentialExpressionService.run_differential_expression_job(cluster_file, study, user, **options)
```
5. In `development.log`, you should see a block for the ingest job launching with all of the parameters, and then continue watching for polling to complete:
```
Ingest run initiated: projects/broad-singlecellportal-bistlin/operations/15071106910722860088, queueing Ingest poller
IngestJob poller: projects/broad-singlecellportal-bistlin/operations/15071106910722860088 is not done; queuing check for 2022-05-09 15:44:54 -0400
IngestJob poller: projects/broad-singlecellportal-bistlin/operations/15071106910722860088 is not done; queuing check for 2022-05-09 15:45:57 -0400
IngestJob poller: projects/broad-singlecellportal-bistlin/operations/15071106910722860088 is not done; queuing check for 2022-05-09 15:47:04 -0400
IngestJob poller: projects/broad-singlecellportal-bistlin/operations/15071106910722860088 is done!
IngestJob poller: projects/broad-singlecellportal-bistlin/operations/15071106910722860088 status: Completed
```
6. Verify that `metadata.is_differential_expression_enabled` has been set:
```
metadata.reload
metadata.is_differential_expression_enabled
=> true
```

This PR satisfies SCP-4195 and SCP-4212.
